digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/LinearAssetService.scala

```diff
@@ -96,23 +98,62 @@ trait LinearAssetOperations {
       rl => rl.linkType.value == UnknownLinkType.value || rl.isCarTrafficRoad)
 
     val timing = System.currentTimeMillis
-    val newAssets = fillNewRoadLinksWithPreviousAssetsData(projectableTargetRoadLinks,
-      existingAssets, assetsOnChangedLinks, changes)
+
+    val (expiredPavingAssetIds, newPavingAssets) = getPavingAssetChanges(existingAssets, roadLinks, changes, typeId)
+
+    val filledNewAssets = fillNewRoadLinksWithPreviousAssetsData(projectableTargetRoadLinks,
+      existingAssets.filterNot(a => expiredPavingAssetIds.exists(_ == a.id)) ++ newPavingAssets, assetsOnChangedLinks, changes)
+
+    val newAssets = newPavingAssets.filterNot(a => filledNewAssets.exists(_ == a.linkId)) ++ filledNewAssets
+
     if (newAssets.nonEmpty) {
       logger.info("Transferred %d assets in %d ms ".format(newAssets.length, System.currentTimeMillis - timing))
     }
-    val groupedAssets = (existingAssets.filterNot(a => newAssets.exists(_.linkId == a.linkId)) ++ newAssets).groupBy(_.linkId)
+    val groupedAssets = (existingAssets.filterNot(a => expiredPavingAssetIds.exists(_ == a.id)).filterNot(a => newAssets.exists(_.linkId == a.linkId)) ++ newAssets).groupBy(_.linkId)
```

Comment:
I am uncertain if this method works with all the cases when assets are transferred and pavement information is updated.